src/client/components/Navigation/LanguageSettings.js

```diff
@@ -0,0 +1,177 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import _ from 'lodash';
+import { Popover } from 'antd';
+import { connect } from 'react-redux';
+import classNames from 'classnames';
+import { availableLocalesToReactIntl } from '../../translations/index';
+import { SUPPORTED_LANGUAGES } from '../../../common/constants/settings';
+import { saveSettings, setLocale } from '../../settings/settingsActions';
+import { getIsAuthenticated, getLocale } from '../../reducers';
+import PopoverMenu, { PopoverMenuItem } from '../PopoverMenu/PopoverMenu';
+import Loading from '../Icon/Loading';
+import './LanguageSettings.less';
+
+@connect(
+  state => ({
+    authenticated: getIsAuthenticated(state),
+    locale: getLocale(state),
+  }),
+  {
+    saveSettings,
+    setLocale,
+  },
+)
+class LanguageSettings extends React.Component {
+  static propTypes = {
+    authenticated: PropTypes.bool.isRequired,
+    saveSettings: PropTypes.func.isRequired,
+    setLocale: PropTypes.func.isRequired,
+    locale: PropTypes.string,
+  };
+
+  static defaultProps = {
+    locale: 'en-US',
+  };
+
+  static getValidLocale = locale => {
+    const availableLocales = _.keys(availableLocalesToReactIntl);
+    let validLocale = 'en-US';
+
+    for (let i = 0; i < availableLocales.length; i += 1) {
+      const currentLocale = availableLocales[i];
+      const currentLocaleShortValue = availableLocalesToReactIntl[currentLocale];
+
+      if (locale === currentLocale) {
+        validLocale = currentLocale;
```

Comment:
I think it would look a little bit easier with returning there instead of using intermediary variable and breaks.